pkg/gohai/utils/common_test.go

```diff
@@ -0,0 +1,259 @@
+	require.ErrorIs(t, err, ErrCannotRender)
+
+	infoValuePtr := &struct {
+		ValuePtr Value[*int] `json:"value_ptr"`
+	}{
+		ValuePtr: NewValue[*int](nil),
+	}
+	_, _, err = AsJSON(infoValuePtr, false)
+	require.ErrorIs(t, err, ErrCannotRender)
+}
+
+func TestAsJsonWarns(t *testing.T) {
+	errs := []string{
+		"this is the first error",
+		"this is the second error",
+		"this is the third error",
+		"this is the fourth error",
+	}
+	info := &struct {
+		FieldOne   Value[int] `json:"field_one"`
+		FieldTwo   Value[int] `json:"field_two"`
+		FieldThree Value[int] `json:"field_three"`
+		FieldFour  Value[int] `json:"field_four"`
+		FieldFive  Value[int] `json:"field_five"`
+	}{
+		FieldOne:   NewErrorValue[int](errors.New(errs[0])),
+		FieldTwo:   NewErrorValue[int](errors.New(errs[1])),
+		FieldThree: NewErrorValue[int](errors.New(errs[2])),
+		FieldFour:  NewErrorValue[int](errors.New(errs[3])),
+		FieldFive:  NewValue(1),
+	}
+
+	marshallable, warns, err := AsJSON(info, false)
+	require.NoError(t, err)
+	require.ElementsMatch(t, errs, warns)
+
+	marshalled, err := json.Marshal(marshallable)
+	require.NoError(t, err)
+
+	expected := `{
+		"field_five": "1"
+	}`
+	require.JSONEq(t, expected, string(marshalled))
+
+	marshallable, warns, err = AsJSON(info, true)
+	require.NoError(t, err)
+	require.ElementsMatch(t, errs, warns)
+
+	marshalled, err = json.Marshal(marshallable)
+	require.NoError(t, err)
+
+	expected = `{
+		"field_one": "0",
+		"field_two": "0",
+		"field_three": "0",
+		"field_four": "0",
+		"field_five": "1"
+	}`
+	require.JSONEq(t, expected, string(marshalled))
+}
+
+func TestAsJSONSuffix(t *testing.T) {
+	errs := []string{
+		"this is the first error",
+		"this is the second error",
+		"this is the third error",
+	}
+	info := &struct {
+		FieldOne   Value[int] `json:"field_one"`
+		FieldTwo   Value[int] `json:"field_two" unit:""`
+		FieldThree Value[int] `json:"field_three" unit:"kb"`
+		FieldFour  Value[int] `json:"field_four"`
+		FieldFive  Value[int] `json:"field_five" unit:""`
+		FieldSix   Value[int] `json:"field_six" unit:"M"`
+	}{
+		FieldOne:   NewValue(1),
+		FieldTwo:   NewValue(2),
+		FieldThree: NewValue(3),
+		FieldFour:  NewErrorValue[int](errors.New(errs[0])),
+		FieldFive:  NewErrorValue[int](errors.New(errs[1])),
+		FieldSix:   NewErrorValue[int](errors.New(errs[2])),
+	}
+
+	marshallable, warns, err := AsJSON(info, false)
+	require.NoError(t, err)
+
+	require.ElementsMatch(t, errs, warns)
+
+	marshalled, err := json.Marshal(marshallable)
+	require.NoError(t, err)
+
+	expected := `{
+		"field_one": "1",
+		"field_two": "2",
+		"field_three": "3kb"
+	}`
+	require.JSONEq(t, expected, string(marshalled))
+
+	marshallable, warns, err = AsJSON(info, true)
+	require.NoError(t, err)
+	require.ElementsMatch(t, errs, warns)
+
+	marshalled, err = json.Marshal(marshallable)
+	require.NoError(t, err)
+
+	expected = `{
+		"field_one": "1",
+		"field_two": "2",
+		"field_three": "3kb",
+		"field_four": "0",
+		"field_five": "0",
```

Comment:
Note: You probably don't need two fields with an error (`field_four` and `field_five`) neither two fields with a value without unit (`field_one`, `field_two`). One for each case can be enough.

Reply:
They are not really the same cases: I'm testing without unit, with empty unit and with a non-empty unit. It's just to make sure the edge case works properly.